### ISSUE TITLE
SnoopDB completion failure

### ISSUE THREAD

**Issue by BobyMCbobs** [cncf/apisnoop#584]
Looking over at
https://github.com/cncf/apisnoop/runs/8179835670?check_suite_focus=true
There appears to be a failure on the loading of the data

```
2022-09-05T04:44:14.0693878Z /usr/local/bin/docker-entrypoint.sh: running /docker-entrypoint-initdb.d/502_load_all_audit_events.sql
2022-09-05T04:44:14.0694363Z BEGIN
2022-09-05T04:44:14.0694751Z 2022-09-04 22:45:46.307 UTC [450] LOG:  our bucket and job
2022-09-05T04:44:14.0695247Z 2022-09-04 22:45:46.307 UTC [450] DETAIL:  ['ci-kubernetes-e2e-gci-gce', '1566545747850039296']
2022-09-05T04:44:14.0695859Z 2022-09-04 22:45:46.307 UTC [450] CONTEXT:  PL/Python function "load_audit_events"
2022-09-05T04:44:14.0696421Z 2022-09-04 22:45:46.307 UTC [450] STATEMENT:  select * from load_audit_events() f("build log");
2022-09-05T04:44:14.0696869Z 2022-09-04 22:45:46.561 UTC [450] LOG:  our finished json
2022-09-05T04:44:14.0697248Z 2022-09-04 22:45:46.561 UTC [450] DETAIL:  {
2022-09-05T04:44:14.0697535Z 		"timestamp": 1662330892,
2022-09-05T04:44:14.0697792Z 		"passed": true,
2022-09-05T04:44:14.0698047Z 		"metadata": {
2022-09-05T04:44:14.0698307Z 			"uploader": "crier"
2022-09-05T04:44:14.0698554Z 		},
2022-09-05T04:44:14.0698785Z 		"result": "success"
2022-09-05T04:44:14.0699027Z 	}
2022-09-05T04:44:14.0699427Z 2022-09-04 22:45:46.561 UTC [450] CONTEXT:  PL/Python function "load_audit_events"
2022-09-05T04:44:14.0699917Z 2022-09-04 22:45:46.561 UTC [450] STATEMENT:  select * from load_audit_events() f("build log");
2022-09-05T04:44:14.0700747Z 2022-09-04 22:45:46.812 UTC [450] ERROR:  KeyError: 'version'
2022-09-05T04:44:14.0701216Z 2022-09-04 22:45:46.812 UTC [450] CONTEXT:  Traceback (most recent call last):
2022-09-05T04:44:14.0701607Z 	  PL/Python function "load_audit_events", line 23, in <module>
2022-09-05T04:44:14.0701981Z 	    auditlog_file = download_and_process_auditlogs(bucket, job)
2022-09-05T04:44:14.0702392Z 	  PL/Python function "load_audit_events", line 347, in download_and_process_auditlogs
2022-09-05T04:44:14.0702767Z 	PL/Python function "load_audit_events"
2022-09-05T04:44:14.0703241Z 2022-09-04 22:45:46.812 UTC [450] STATEMENT:  select * from load_audit_events() f("build log");
2022-09-05T04:44:19.1311026Z ##[error]The operation was canceled.
```

**Comment by zachmandeville** [comment]
The output of the finished json looks strange to me.
We are trying to pull the metadata for the bucket `ci-kubernetes-e2e-gci-gce`.
For the last successfully parsed job (up now on apisnoop) that metadata looks like this:
```
{
  "timestamp": 1661600166, 
  "version": "v1.26.0-alpha.0.275+41df8167dd82e7", 
  "result": "SUCCESS", 
  "passed": true, 
  "job-version": "v1.26.0-alpha.0.275+41df8167dd82e7", 
  "metadata": {
    "kubetest-version": "v20220822-9f275332ba", 
    "infra-commit": "6bf2155b8", 
    "worker_node_os_image": "cos-97-16919-103-16", 
    "job-version": "v1.26.0-alpha.0.275+41df8167dd82e7", 
    "control_plane_node_os_image": "cos-97-16919-103-16", 
    "revision": "v1.26.0-alpha.0.275+41df8167dd82e7"
  }
}
```

Now, we are trying to pull metadata for `ci-kubernetes-e2e-gci-gce/1566545747850039296` and it gives us this:
```
{
	"timestamp": 1662478802,
	"passed": true,
	"metadata": {
		"uploader": "crier"
	},
	"result": "success"
}
```
This ultimately is the cause of the error.

## The error

in our [download_and_process_audit_logs function](https://github.com/cncf/apisnoop/blob/6a012c3a009f607a5b8a5d16cc5c6cd356beb319/apps/snoopdb/postgres/snoopUtils.py#L320), we parse the 
metadata's version key to get the commit of kubernetes used in the job:
```
commit_hash = metadata["version"].split("+")[1]
```
So in our working job, it's metadata["version"] is `v1.26.0-alpha.0.275+41df8167dd82e7` and so the above line would 
give us the commit hash `41df8167dd82e7`

In this current job, there is no metadata["version"] and so it throws an error.

A "fix" would be to attempt to grab the key and return something else (like 'master') if it can't be found.

The problem is that we use the commit hash to grab the exact version of the kubernetes swagger.json we need, so that we can map the audit event entries to their appropriate operation ID's (or endpoints).  If we are  pulling from 'master' it may be out-of-date compared to the actual test run, and so our coverage report would be inaccurate.

In this case, I think a big error and a failed job is the right way to go, instead of a quiet failure and inaccurate results.

## What Next?

I am uncertain why the metadata looks like that. Is this a new form for particular test runs, or just an anomaly?  

We determine which test run to grab by going to the bucket's [jobResultsCache.json](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/jobResultsCache.json) and pulling the topmost successful test run.  This means that, when this gh action was run, job `1566545747850039296` showed up as the latest successful test. Looking at its [spyglass page](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce/1566545747850039296), it does look like it ran through the conformance tests successfully and has the right artifacts.  The only issue is its `finished.json` is different.

I see two options for us: 
1.) Investigate why this file changed its format, and if we can expect it to happen again. Lobby with community to make sure this file structure is consistent.
2.) Find a better way of getting the commit hash from the artifacts of a job.  Determine which bit of metadata will both tell us the commit hash _and_ be guaranteed a consistent structure.

I think the second option is a better one.  I'll keep investigating, and would love to discuss if there's a better way for us to determine the `k8s/k8s` commit hash of a test run.

**Comment by zachmandeville** [comment]
Investigated this further and haven't found a satisfying solution.  I'll try to summarize where I'm at:

- APISnoop is built off taking the conformance test suite, and all the endpoints it hits, and calculating coverage.
- This means we need some storage object that contains all the test suite logs.
- They are not held in a single storage object, but three different ones.
- For each one, we want to grab the latest results, when they ran, and which version of k8s they tested.
- Two of the buckets give us this through a jobResultsCache.json, the third one does not.
- for that one we have to do a webscrape of it's test page just to get the latest successful job, but it won't tell us when it ran or its k8s version.
- We decided to use a finished.json that contains both info for all three.
- This finished.json changed for 1 of the buckets so it no longer has the info we need.
- Two of the buckets give us the info in a different JSON file, but the third one does not contain that json file.
- We now have three different buckets and 2-3 different ways we need to follow to get the info we want.
- We cannot trust that this will stay stable, it seems.
- I think we need to do a revisit of how we are trying to get this info.

**Comment by zachmandeville** [comment]
This should be fixed with #588

**Comment by BobyMCbobs** [comment]
> This should be fixed with #588

Thank you @zachmandeville!
